### PR TITLE
Added delayed shutdown feature

### PR DESCRIPTION
**Pull request by pinkysnowman**

### CODE REVIEW

File: builtin/game/chatcommands.lua [hunk 1]
```diff
@@ -774,12 +774,28 @@ core.register_chatcommand("time", {
...
+			if param > 0 or param < 360 then
+				core.log("action", name .. " shuts down server, delayed "..param.." seconds.")
+				core.chat_send_all("*** Server will be shutting down in "..param.." seconds (operator request).")
+				core.after(param, function()
+				core.chat_send_all("*** Server shutting down (operator request).")
+				core.request_shutdown()
+				end)
+				return true, "shutdowning server down in "..time.." seconds!!!"
+			else
+				return false, "Delay time must be between 1 and 3600"
```
Comment (cheapie) [inline builtin/game/chatcommands.lua]: Is it 360 or 3600?

File: builtin/game/chatcommands.lua [hunk 2]
```diff
@@ -774,12 +774,28 @@ core.register_chatcommand("time", {
...
+	params = "<delay_in_seconds(0..360)> | Leave blank for normal shutdown",
 	description = "shutdown server",
 	privs = {server=true},
 	func = function(name, param)
-		core.log("action", name .. " shuts down server")
-		core.request_shutdown()
-		core.chat_send_all("*** Server shutting down (operator request).")
+		if param ~= "" then
+			param = tonumber(param) or 0
+			if param > 0 or param < 360 then
```
Comment (cheapie) [inline builtin/game/chatcommands.lua]: So... all real numbers? It's great that I can't shut down the server in 72+3i seconds, at least...

/shutdown -1

*crash*

Well, it does shut it down.

File: builtin/game/chatcommands.lua [hunk 3]
```diff
@@ -774,12 +774,28 @@ core.register_chatcommand("time", {
...
+				core.log("action", name .. " shuts down server, delayed "..param.." seconds.")
+				core.chat_send_all("*** Server will be shutting down in "..param.." seconds (operator request).")
+				core.after(param, function()
+				core.chat_send_all("*** Server shutting down (operator request).")
+				core.request_shutdown()
+				end)
+				return true, "shutdowning server down in "..time.." seconds!!!"
```
Comment (cheapie) [inline builtin/game/chatcommands.lua]: I don't think "shutdowning" is a word.

**Comment by cheapie** [comment]
Does not work (crashes server).

See diff comments.

**Comment by rubenwardy** [comment]
Why do you limit to 6 minutes?

File: builtin/game/chatcommands.lua (hunk 3 above)
Comment (kaeza) [inline builtin/game/chatcommands.lua]: Also, no "!!!" please.

File: builtin/game/chatcommands.lua [hunk 4]
```diff
@@ -774,12 +774,28 @@ core.register_chatcommand("time", {
...
-		core.log("action", name .. " shuts down server")
-		core.request_shutdown()
-		core.chat_send_all("*** Server shutting down (operator request).")
+		if param ~= "" then
+			param = tonumber(param) or 0
+			if param > 0 or param < 360 then
+				core.log("action", name .. " shuts down server, delayed "..param.." seconds.")
+				core.chat_send_all("*** Server will be shutting down in "..param.." seconds (operator request).")
+				core.after(param, function()
+				core.chat_send_all("*** Server shutting down (operator request).")
```
Comment (kaeza) [inline builtin/game/chatcommands.lua]: Indentation is broken here.

Comment (pinkysnowman) [inline builtin/game/chatcommands.lua]: Every time, my editor kills me on github.  Will correct both things :)

**Comment by pinkysnowman** [comment]
@cheapie My english is crappy i know, and it was origonally 3600 but was reduced later.
@rubenwardy this is more to prevent an operator from typoing or setting it so far out they forget about it
@kaeza fixed both :)

the origional cone was from one of my mods but i removed some of it as i felt it to be too bulky to do a PR on. It did warn via chat_send_all every 10 seconds as a countdown, part of the reason it was limited to 360 seconds!

**Comment by kaeza** [comment]
As rubenwardy, I'm not so sure about the 6 min limit. Would like to hear more opinions.

Otherwise, :+1:.

**Comment by rubenwardy** [comment]
I think the limit should be at least a day. When the shutdown is scheduled, it should say "Shutting down in 6h5m3s" or something. It would also be good to be able to accept, instead of seconds, something like 6h3m.

It should also be possible to give a shutdown message. That's out of scope for this PR, though, but should be taken into account.

You should also add a cancel_shutdown command.

**Comment by kaeza** [comment]
>Xban2 does that.

[Here](https://github.com/kaeza/minetest-xban2/blob/master/init.lua#L31) it is, if wanted. Too trivial for copyright purposes, so feel free to use. (Attribution appreciated but not required)

**Comment by HybridDog** [comment]
You know that minetest.after doesn't work right, don't you?
https://github.com/minetest/minetest/pull/3443

**Comment by pinkysnowman** [comment]
I have never had any issues with it

**Comment by pinkysnowman** [comment]
the only issue with a timer is once its in (ex: 6h 20m) in that amount of time the server will shut down, EVEN AFTER A CRASH OR RESTART! as this may cause major issues it is the origional reason i limited it to 6 minuites :)

**Comment by rubenwardy** [comment]
> the only issue with a timer is once its in (ex: 6h 20m) in that amount of time the server will shut down, EVEN AFTER A CRASH OR RESTART! as this may cause major issues it is the origional reason i limited it to 6 minuites :)

Nope. Minetest.after isn't persisted: if the server shuts down before minetest.after is run then minetest.after will never run.

**Comment by pinkysnowman** [comment]
are you sure, because i have gotten it shutdown in that way after a crash like it is or on an older build was. :/

**Comment by rubenwardy** [comment]
Pretty sure. They're not saved anywhere here: https://github.com/minetest/minetest/blob/master/builtin/game/misc.lua#L7-L75

**Comment by nerzhul** [comment]
This feature is interesting, i will backport a change from epixel to minetest because my delayed shutdown feature is a little bit more complete, this PR doesn't permit to cancel, don't notify connecting players, etc...

**Comment by nerzhul** [comment]
See #4664

**Pull request closed without merging**